File: today2/linked-list-cycle-ii.ts
/**
 * https://leetcode.com/problems/linked-list-cycle-ii/description/
 */
export { detectCycle, ListNode };

function detectCycle(head: ListNode | null): ListNode | null {
  return null;
}

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

File: python/today11/linked_list_cycle_ii.py
# https://leetcode.com/problems/linked-list-cycle-ii/description/
from typing import Optional

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        
        cache = set()

        while head.next:
            if head.next in cache:
                return head.next
            cache.add(head)
            head = head.next

        return None
